docs/Compiler_Hardening_Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md

```diff
@@ -245,26 +245,42 @@ However, when enabling `_FORTIFY_SOURCE=2` in existing code bases regression tes
+| <span id="-D_LIBCPP_ASSERT">`-D_LIBCPP_ASSERT`</span>                                      | libc++ 3.3.0       | (C++ using libc++ only) Constant-time precondition checks for libc++ calls. |
 
 #### Synopsis
 
-The `-D_GLIBCXX_ASSERTIONS` macro enables run-time bounds checking for C++ strings and containers. It can only affect C++ code.
+The C++ standard library implementations in GCC (libstdc++) and LLVM (libc++) provide run-time precondition checks for C++ standard library calls, such as bounds-checks for C++ strings and containers, and null-pointer checks when dereferencing smart pointers.
+
+These precondition checks can be enabled by defining the corresponding pre-processor macros in when compiling C++ code that calls into libstdc++ or libc++:
+
+- The `-D_GLIBCXX_ASSERTIONS` macro enables precondition checks for libstdc++[^libstdc++_macros].  
+  It can only affect C++ code that uses GCC’s libstdc++.
+- The `-D_LIBCPP_ASSERT` macro enables precondition checks for libc++[^Clow19].  
+  It can only affect C++ code that uses LLVM’s libc++.
 
 #### Performance implications
 
+Most calls into the C++ standard library have preconditions. Some preconditions can be checked in constant-time, others are more expensive.
+Both `-D_GLIBCXX_ASSERTIONS` and `-D_LIBCPP_ASSERT` are intended to enable only lightweight[^Wakely15], i.e., constant-time checks[^Dionne22] but the exact behavior can differ between standard library versions .
+
 The `-D_GLIBCXX_ASSERTIONS` macro can have a non-trivial impact on performance.
 Impacts of [up to 6% on performance have been reported](https://gitlab.psi.ch/OPAL/src/-/merge_requests/468).
 
 #### When not to use?
 
-`-D_GLIBCXX_ASSERTIONS` is recommended for C++ applications that may handle untrusted data, as well as for any C++ application during testing.
+`-D_GLIBCXX_ASSERTIONS` and `-D_LIBCPP_ASSERT` are recommended for C++ applications that may handle untrusted data, as well as for any C++ application during testing.
+
+These options are unnecessary for security for applications in production that only handle completely trusted data.
 
-Option `-D_GLIBCXX_ASSERTIONS` is unnecessary for security for applications in production that only handle completely trusted data.
+[^libstdc++_macros]: The GNU C++ Library: Using Macros. https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_macros.html
+[^Clow19]: Marshall Clow. Hardening the C++ standard template library. C++ Russia 2019. https://www.youtube.com/watch?v=1iHs_K2HpGo&t=990s
+[^Wakely15]: Jonathan Wakely. Enable lightweight checks with _GLIBCXX_ASSERTIONS. GCC Mailing List, Sept. 7, 2015. https://patchwork.ozlabs.org/project/gcc/patch/20150907182755.GP2631@redhat.com/
+[^Dionne22]: Loius Dionne. Audit all uses of _LIBCPP_ASSERT and _LIBCPP_DEBUG_ASSERT. https://github.com/llvm/llvm-project/commit/c87c8917e3662532f0aa75a91caea857c093f8f4
```

Comment:
should these references live closer to their relevant lines?

---

There doesn't seem to be much standardization around referencing in the existing doc:

```

[^Cimpanu2019]: Cimpanu, Catalin, [Microsoft: 70 percent of all security bugs are memory safety issues](https://www.zdnet.com/article/microsoft-70-percent-of-all-security-bugs-are-memory-safety-issues/), ZDNet, 2019-02-11
[^Cimpanu2020]: Cimpanu, Catalin, [Chrome: 70% of all security bugs are memory safety issues](https://www.zdnet.com/article/chrome-70-of-all-security-bugs-are-memory-safety-issues/), ZDNet, 2020-05-22
[^1]: The implementation of `-D_FORTIFY_SOURCE={1,2}` in the GNU libc (glibc) relies heavily on implementation details within GCC. Clang implements its own style of fortified function calls (originally introduced for Android’s bionic libc) and but as of Clang / LLVM 14.0.6 incorrectly produces non-fortified calls to glibc functions with `_FORTIFY_SOURCE` . Code set to be fortified with Clang will still compile does not benefit from the fortified function variants in glibc. For more information see: Guelton, Serge. Toward _FORTIFY_SOURCE parity between Clang and GCC. Red Hat Developer.
 <https://developers.redhat.com/blog/2020/02/11/toward-_fortify_source-parity-between-clang-and-gcc> and
 Poyarekar, Siddhesh. D91677 Avoid simplification of library functions when callee has an implementation. (LLVM Phabricator) <https://reviews.llvm.org/D91677>
[^3]: Clang documentation: Diagnostics flags in Clang.
<https://clang.llvm.org/docs/DiagnosticsReference.html>
```

But this feels _maybe_ a bit more like what currently exists:

```suggestion
[^libstdc++_macros]: The GNU C++ Library: Using Macros.
<https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_macros.html>
[^Clow19]: Clow, Marshall - [Hardening the C++ standard template library](https://www.youtube.com/watch?v=1iHs_K2HpGo&t=990s), C++ Russia 2019. 
[^Wakely15]: Wakely, Jonathan - Enable lightweight checks with _GLIBCXX_ASSERTIONS. GCC Mailing List, 2015-09-07.
<https://patchwork.ozlabs.org/project/gcc/patch/20150907182755.GP2631@redhat.com/>
[^Dionne22]: Dionne, Loius - [Audit all uses of _LIBCPP_ASSERT and _LIBCPP_DEBUG_ASSERT](https://github.com/llvm/llvm-project/commit/c87c8917e3662532f0aa75a91caea857c093f8f4)
```

I swapped out the comma after the first names though because that looks awful and confusing IMO :sweat_smile:

Ozlabs Patchwork is a nice interface to the mailing list as it shows replies but maybe it's best to use original links?

https://gcc.gnu.org/pipermail/gcc-patches/2015-September/427766.html